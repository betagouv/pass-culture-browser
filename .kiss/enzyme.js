import React from 'react'
import { shallow } from 'enzyme'

import MyComponent from '../MyComponent'

describe('src | components | MyComponent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const props = { errors: {} }
      const wrapper = shallow(<MyComponent {...props} />)
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
