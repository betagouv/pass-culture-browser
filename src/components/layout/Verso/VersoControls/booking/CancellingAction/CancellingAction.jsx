import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

import Price from '../../../../Price/Price'
import getIsCancelling from '../../../../../../helpers/getIsCancelling'

class CancellingAction extends PureComponent {
  componentDidMount() {
    const { booking, match, offer, openCancelPopin } = this.props
    const { id: bookingId } = booking
    const { name: offerName, id: offerId } = offer
    const isCancelling = getIsCancelling(match)

    if (isCancelling) {
      openCancelPopin(bookingId, offerName, offerId)
    }
  }

  componentDidUpdate(prevProps) {
    const { booking, match, offer, openCancelPopin } = this.props
    const { id: bookingId } = booking
    const { name: offerName, id: offerId } = offer
    const isCancelling = getIsCancelling(match)
    const previousIsCancelling = getIsCancelling(prevProps.match)

    if (isCancelling && !previousIsCancelling) {
      openCancelPopin(bookingId, offerName, offerId)
    }
  }

  handleCancellingAction = () => {
    const { cancellingUrl, history } = this.props

    history.push(cancellingUrl)
  }

  render() {
    const { price } = this.props

    return (
      <button
        className="ticket-action"
        onClick={this.handleCancellingAction}
        type="button"
      >
        <span className="ticket-price ticket-reserved">
          <Price
            free="Gratuit"
            value={price}
          />
          <i className="icon-ico-check ticket-reserved-icon" />
        </span>
        <span className="ticket-label">
          {'Annuler'}
        </span>
      </button>
    )
  }
}

CancellingAction.propTypes = {
  booking: PropTypes.shape().isRequired,
  cancellingUrl: PropTypes.string.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      bookings: PropTypes.string,
      bookingId: PropTypes.string,
      cancellation: PropTypes.string,
    }).isRequired,
  }).isRequired,
  offer: PropTypes.shape().isRequired,
  openCancelPopin: PropTypes.func.isRequired,
  price: PropTypes.number.isRequired,
}

export default CancellingAction
