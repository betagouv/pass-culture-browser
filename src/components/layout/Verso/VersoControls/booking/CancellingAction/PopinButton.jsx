import PropTypes from 'prop-types'
import React from 'react'

const onClick = action => () => {
  document.querySelector('.popin-button').disabled = true
  action()
}

const PopinButton = ({ action, label }) => (
  <button
    className="popin-button"
    key={label}
    onClick={onClick(action)}
    type="button"
  >
    {label}
  </button>
)

PopinButton.propTypes = {
  action: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
}

export default PopinButton
