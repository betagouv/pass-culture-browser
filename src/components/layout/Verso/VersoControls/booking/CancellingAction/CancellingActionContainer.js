import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'
import { requestData } from 'redux-thunk-data'

import withTracking from '../../../../../hocs/withTracking'
import CancellingAction from './CancellingAction'
import PopinButton from './PopinButton'
import { bookingNormalizer } from '../../../../../../utils/normalizers'
import { closeSharePopin, openSharePopin } from '../../../../../../reducers/share'
import { selectBookingByRouterMatch } from '../../../../../../selectors/data/bookingsSelector'
import { selectOfferByRouterMatch } from '../../../../../../selectors/data/offersSelector'
import selectStockById from '../../../../../../selectors/selectStockById'

export const getCancellingUrl = (bookingId, params, pathname, search) => {
  let bookingUrl = pathname

  bookingUrl = `${bookingUrl}/reservation`

  if (params.bookingId === undefined) {
    bookingUrl = `${bookingUrl}/${bookingId}`
  }

  return `${bookingUrl}/annulation${search}`
}

export const mapStateToProps = (state, ownProps) => {
  const { location, match } = ownProps
  const { pathname, search } = location
  const booking = selectBookingByRouterMatch(state, match)
  const cancellingUrl = getCancellingUrl(booking.id, match.params, pathname, search)
  const offer = selectOfferByRouterMatch(state, match)
  const stock = selectStockById(state, booking.stockId)
  const price = stock.price * booking.quantity

  return { booking, cancellingUrl, offer, price }
}

export const mapDispatchToProps = (dispatch, ownProps) => {
  const { history, location } = ownProps
  const { pathname, search } = location

  const handleClosePopin = () => {
    dispatch(closeSharePopin())
    const nextPathname = pathname.split(/\/reservation(\/|$|\/$)/)[0]
    const nextUrl = `${nextPathname}${search}`
    history.push(nextUrl)
  }

  const handleOpenFailurePopin = (state, request) => {
    const { payload } = request
    const { errors } = payload
    const { booking: bookingError } = errors || {}
    const message = bookingError || ['Une erreur inconnue s’est produite']

    const options = {
      buttons: [
        PopinButton({
          action: handleClosePopin,
          label: 'OK',
        }),
      ],
      text: message.join('\n'),
      title: 'Annulation impossible',
    }
    dispatch(openSharePopin(options))
  }

  const handleSuccessPopin = offerId => {
    dispatch(closeSharePopin())
    const successUrl = `${pathname}/confirmation${search}`
    ownProps.tracking.trackEvent({ action: 'cancelBooking', name: offerId })
    history.push(successUrl)
  }

  const requestPatchBooking = (bookingId, offerId) => {
    dispatch(
      requestData({
        apiPath: `/bookings/${bookingId}`,
        body: { isCancelled: true },
        handleFail: handleOpenFailurePopin,
        handleSuccess: () => handleSuccessPopin(offerId),
        method: 'PATCH',
        normalizer: bookingNormalizer,
      })
    )
  }

  return {
    openCancelPopin: (bookingId, offerName, offerId) => {
      const options = {
        buttons: [
          PopinButton({
            action: () => requestPatchBooking(bookingId, offerId),
            label: 'Oui',
          }),
          PopinButton({
            action: handleClosePopin,
            label: 'Non',
          }),
        ],
        handleClose: () => history.push(`${pathname.split('/reservation/')[0]}${search}`),
        text: 'Souhaitez-vous réellement annuler cette réservation ?',
        title: offerName,
      }
      dispatch(openSharePopin(options))
    },
  }
}

export default compose(
  withRouter,
  withTracking('Offer'),
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(CancellingAction)
