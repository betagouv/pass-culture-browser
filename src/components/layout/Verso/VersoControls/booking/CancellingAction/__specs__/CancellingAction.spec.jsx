import { mount, shallow } from 'enzyme'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Provider } from 'react-redux'
import { Route, Router } from 'react-router'

import CancellingAction from '../CancellingAction'
import CancellingActionContainer from '../CancellingActionContainer'
import getMockStore from '../../../../../../../utils/mockStore'

jest.mock('redux-thunk-data', () => {
  const { requestData } = jest.requireActual('fetch-normalize-data')

  return {
    requestData,
  }
})

describe('src | components | layout | Verso | VersoControls | booking | CancellingAction | CancellingAction', () => {
  let props

  beforeEach(() => {
    props = {
      booking: {
        id: 'AAA',
        recommendation: {
          offerId: 'BBB',
        },
      },
      cancellingUrl: '/reservation/annulation',
      history: {
        push: jest.fn(),
      },
      location: {
        pathname: '',
        search: '',
      },
      match: { params: {} },
      offer: {},
      openCancelPopin: jest.fn(),
      price: 20,
    }
  })

  describe('when the offer is booking', () => {
    it('should render a button to cancelling the booking', () => {
      // when
      const wrapper = shallow(<CancellingAction {...props} />)

      // then
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when I click on button for cancelling', () => {
    it('should change the pathname', () => {
      // given
      const mockHistory = createMemoryHistory()
      mockHistory.push('/reservations/details/b1?param=value')
      const mockStore = getMockStore({
        data: (
          state = {
            bookings: [
              {
                id: 'b1',
                stockId: 's1',
              },
            ],
            mediations: [],
            offers: [
              {
                id: 'o1',
              },
            ],
            stocks: [
              {
                id: 's1',
                offerId: 'o1',
                price: 20,
              },
            ],
          }
        ) => state,
      })
      const wrapper = mount(
        <Provider store={mockStore}>
          <Router history={mockHistory}>
            <Route path="/reservations/:details(details|transition)?/:bookingId?/:booking(reservation)?/:cancellation(annulation)?/:confirmation(confirmation)?/:menu(menu)?">
              <CancellingActionContainer />
            </Route>
          </Router>
        </Provider>
      )

      // when
      wrapper.find({ children: 'Annuler' }).simulate('click')

      // then
      expect(`${mockHistory.location.pathname}${mockHistory.location.search}`).toBe(
        '/reservations/details/b1/reservation/annulation?param=value'
      )
    })
  })
})
