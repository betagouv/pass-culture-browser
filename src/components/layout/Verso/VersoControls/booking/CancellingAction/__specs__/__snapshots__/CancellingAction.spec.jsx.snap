// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | Verso | VersoControls | booking | CancellingAction | CancellingAction when the offer is booking should render a button to cancelling the booking 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CancellingAction
    booking={
      Object {
        "id": "AAA",
        "recommendation": Object {
          "offerId": "BBB",
        },
      }
    }
    cancellingUrl="/reservation/annulation"
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={
      Object {
        "pathname": "",
        "search": "",
      }
    }
    match={
      Object {
        "params": Object {},
      }
    }
    offer={Object {}}
    openCancelPopin={[MockFunction]}
    price={20}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="ticket-price ticket-reserved"
        >
          <Price
            className=""
            free="Gratuit"
            value={20}
          />
          <Icon
            alt=""
            className={null}
            png={null}
            svg="ico-check-w"
          />
        </span>,
        <span
          className="ticket-label"
        >
          Annuler
        </span>,
      ],
      "className": "ticket-action",
      "onClick": [Function],
      "type": "button",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Price
              className=""
              free="Gratuit"
              value={20}
            />,
            <Icon
              alt=""
              className={null}
              png={null}
              svg="ico-check-w"
            />,
          ],
          "className": "ticket-price ticket-reserved",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "free": "Gratuit",
              "value": 20,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "alt": "",
              "className": null,
              "png": null,
              "svg": "ico-check-w",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Annuler",
          "className": "ticket-label",
        },
        "ref": null,
        "rendered": "Annuler",
        "type": "span",
      },
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="ticket-price ticket-reserved"
          >
            <Price
              className=""
              free="Gratuit"
              value={20}
            />
            <Icon
              alt=""
              className={null}
              png={null}
              svg="ico-check-w"
            />
          </span>,
          <span
            className="ticket-label"
          >
            Annuler
          </span>,
        ],
        "className": "ticket-action",
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Price
                className=""
                free="Gratuit"
                value={20}
              />,
              <Icon
                alt=""
                className={null}
                png={null}
                svg="ico-check-w"
              />,
            ],
            "className": "ticket-price ticket-reserved",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "free": "Gratuit",
                "value": 20,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "",
                "className": null,
                "png": null,
                "svg": "ico-check-w",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Annuler",
            "className": "ticket-label",
          },
          "ref": null,
          "rendered": "Annuler",
          "type": "span",
        },
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
